text/0000-impl-specialization.md

```diff
@@ -0,0 +1,1652 @@
+}
+
+fn weaponize() -> bool {
+    let b: Box<bool> = trouble(true);
+    *b
+}
+```
+
+What's going on? When type checking `trouble`, the compiler has a type `T` about
+which it knows nothing, and sees an attempt to employ the `Example` trait via
+`Example::generate(t)`. Because of the blanket impl, this use of `Example` is
+allowed -- but furthermore, the associated type found in the blanket impl is now
+directly usable, so that `<T as Example>::Output` is known within `trouble` to
+be `Box<T>`, allowing `trouble` to type check. But during *monomorphization*,
+`weaponize` will actually produce a version of the code that returns a boolean
+instead, and then attempt to dereference that boolean. In other words, things
+look different to the typechecker than they do to codegen. Oops.
+
+So what went wrong? It should be fine for the compiler to assume that `T:
+Example` for all `T`, given the blanket impl. But it's clearly problematic to
+*also* assume that the associated types will be the ones given by that blanket
+impl. Thus, the "obvious" solution is just to generate a type error in `trouble`
+by preventing it from assuming `<T as Example>::Output` is `Box<T>`.
+
+Unfortunately, this solution doesn't work. For one thing, it would be a breaking
+change, since the following code *does* compile today:
+
+```rust
+trait Example {
+    type Output;
+    fn generate(self) -> Self::Output;
+}
+
+impl<T> Example for T {
+    type Output = Box<T>;
+    fn generate(self) -> Box<T> { Box::new(self) }
+}
+
+fn trouble<T>(t: T) -> Box<T> {
+    Example::generate(t)
+}
+```
+
+And there are definitely cases where this pattern is important. To pick just one
+example, consider the following impl for the slice iterator:
+
+```rust
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+    // ...
+}
+```
+
+It's essential that downstream code be able to assume that `<Iter<'a, T> as
+Iterator>::Item` is just `&'a T`, no matter what `'a` and `T` happen to be.
+
+Furthermore, it doesn't work to say that the compiler can make this kind of
+assumption *unless* specialization is being used, since we want to allow
+downstream crates to add specialized impls. We need to know up front.
+
+Another possibility would be to simply disallow specialization of associated
+types. But the trouble described above isn't limited to associated types. Every
+function/method in a trait has an implicit associated type that implements the
+closure types, and similar bad assumptions about blanket impls can crop up
+there. It's not entirely clear whether they can be weaponized, however. (That
+said, it may be reasonable to stabilize only specialization of functions/methods
+to begin with, and wait for strong use cases of associated type specialization
+to emerge before stabilizing that.)
+
+The solution proposed in this RFC is instead to treat specialization of items in
+a trait as a per-item *opt in*, described in the next section.
+
+## The `default` keyword
```

Comment:
This mentions that `default` will be a keyword, but we currently have modules like `std::default` and methods like `Default::default`, so I think adding this as a keyword may be a breaking change (same with `partial` below). Could this RFC perhaps clarify that they'll be contextual keywords? I believe that should be backwards compatible, right?